keycloak/authentication_subflow.go

```diff
@@ -0,0 +1,162 @@
+	Priority      int    `json:"-"`
+	Requirement   string `json:"-"`
+}
+
+//each subflow creates a flow and an execution under the covers
+type authenticationSubFlowCreate struct {
+	Alias       string `json:"alias"`
+	Type        string `json:"type"`     //providerId of the flow
+	Provider    string `json:"provider"` //authenticator of the execution
+	Description string `json:"description"`
+}
+
+func (keycloakClient *KeycloakClient) NewAuthenticationSubFlow(authenticationSubFlow *AuthenticationSubFlow) error {
+	authenticationSubFlow.TopLevel = false
+	authenticationSubFlow.BuiltIn = false
+	authenticationSubFlowCreate := &authenticationSubFlowCreate{
+		Alias:       authenticationSubFlow.Alias,
+		Type:        authenticationSubFlow.ProviderId,    //providerId of the flow
+		Provider:    authenticationSubFlow.Authenticator, //seems this can be empty //authenticator of the execution
+		Description: authenticationSubFlow.Description,
+	}
+
+	_, location, err := keycloakClient.post(fmt.Sprintf("/realms/%s/authentication/flows/%s/executions/flow", authenticationSubFlow.RealmId, authenticationSubFlow.ParentFlowAlias), authenticationSubFlowCreate)
+	if err != nil {
+		return err
+	}
+	authenticationSubFlow.Id = getIdFromLocationHeader(location)
+
+	if authenticationSubFlow.Requirement != "DISABLED" {
+		return keycloakClient.UpdateAuthenticationSubFlow(authenticationSubFlow)
+	}
+	return nil
+}
+
+func (keycloakClient *KeycloakClient) GetAuthenticationSubFlow(realmId, parentFlowAlias, id string) (*AuthenticationSubFlow, error) {
+	var authenticationSubFlow AuthenticationSubFlow
+	err := keycloakClient.get(fmt.Sprintf("/realms/%s/authentication/flows/%s", realmId, id), &authenticationSubFlow, nil)
+	if err != nil {
+		return nil, err
+	}
+	authenticationSubFlow.RealmId = realmId
+	authenticationSubFlow.ParentFlowAlias = parentFlowAlias
+
+	executionId, err := keycloakClient.getExecutionId(&authenticationSubFlow)
+	if err != nil {
+		return nil, err
+	}
+
+	subFlowExecution, err := keycloakClient.GetAuthenticationExecution(realmId, parentFlowAlias, executionId)
+	if err != nil {
+		return nil, err
+	}
+	authenticationSubFlow.Authenticator = subFlowExecution.Authenticator
+	authenticationSubFlow.Requirement = subFlowExecution.Requirement
+
+	return &authenticationSubFlow, nil
+}
+
+func (keycloakClient *KeycloakClient) getExecutionId(authenticationSubFlow *AuthenticationSubFlow) (string, error) {
+	list, err := keycloakClient.ListAuthenticationExecutions(authenticationSubFlow.RealmId, authenticationSubFlow.ParentFlowAlias)
+	if err != nil {
+		return "", err
+	}
+
+	var executionId = ""
+	for _, ex := range list {
+		if ex.FlowId == authenticationSubFlow.Id {
+			executionId = ex.Id
+			break
+		}
+	}
+	if executionId == "" {
+		return "", errors.New("no execution id found for subflow")
+	}
+	return executionId, nil
```

Comment:
if you return in the for loop, then you can assume this fails if the code reaches this point